ipython_suggestions.py

```diff
@@ -289,29 +335,24 @@ def inspect_all_objs():
                                 objs[name][('module', modulepath)] = (filepath, 0)
 
                                 try:
-                                    with open(filepath, 'r') as f:
-                                        for i, line in enumerate(f):
-                                            if modulepath:
-                                                fullpath = '%s.%s' % (modulepath, name)
-                                            else:
-                                                fullpath = name
-
-                                            m = defclass.match(line)
-                                            if m:
-                                                t, sym = m.groups()
-                                                objs[sym][(t, fullpath)] = (filepath, i)
-                                            else:
-                                                m = variable.match(line)
-                                                if m:
-                                                    objs[m.group(1)][('var', fullpath)] = (
-                                                        filepath, i)
+                                    new_objs = cached_parse_file(filepath, modulepath, name)
+                                    for key, d in new_objs.items():
+                                        objs[key].update(d)
                                 except:
                                     pass
 
         for word, value in objs.items():
             _symbols_cache[len(word)][word] = value
 
         _symbols_sorted = sorted(sum(map(list, _symbols_cache.values()), []))
+        try:
+            import pickle
+            _dumps = pickle.dumps(_global_foo_cache)
+            with open(_cache_file, "wb") as fout:
+                # separate write() from dump() to minimize the probability of a race-condition
+                fout.write(_dumps)
```

Comment:
To prevent accumulation of old entries, one can evict old entries before dumping